### PR TITLE
feat: Added the first iteration of trend insghts breakdowns

### PR DESCRIPTION
**Pull request by Gilbert09**
## Problem
Continue translating the trends insights to use HogQL from https://github.com/PostHog/meta/issues/130

## Changes
- Adding support for being able to breakdown a trends series by an events properties (e.g. device id, or browser)
- Pulled out the query-building logic from the main query runner and into it's own class
- Only supports `string` based event breakdown properties right now, other types to come next

Screenshot 2023-10-10 at 14 42 30

## How did you test this code?
By clicking around in localhost.. testing to be added once full breakdown support is finished

### CODE REVIEW

**Review by mariusandra — COMMENTED** [review]
I'll dig deeper in a bit... however seeing that you split out a `query_builder.py`, one consideration to keep in mind is that we should find a way to have a `to_persons_query()` method somewhere in there. This method return a special version of the same query, that just returns one column: `distinct person_id`.

This will be used to power the new "persons modal", which opens when you click on any datapoint to drill deeper.

We have a very simple version of this working for lifecycles: https://github.com/PostHog/posthog/blob/7dfab45fdb466c408c46f0a802e43aac4457232c/posthog/hogql_queries/insights/lifecycle_query_runner.py#L75-L86 ... though still without any breakdown support (e.g. you don't see all `returning` users from Tuesday, but all users from all the time). 

Without going deep, I'm now not sure if that's an easy to add extra, or something that might require a rethink of the builder.

File: posthog/hogql_queries/insights/trends/breakdown_values.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown_values.py:45]: We don't need the `team_id` field in HogQL, it gets added automatically. Seems like this placeholder is not in use here anyway.

File: posthog/hogql_queries/insights/trends/breakdown_values.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown_values.py:63]: Same here, this can be omitted.

**Review by mariusandra — COMMENTED** [review]
quick comment on one thing before I run to a meeting. Will review more after.

File: posthog/hogql_queries/insights/trends/breakdown_values.py
Comment (Gilbert09) [reply posthog/hogql_queries/insights/trends/breakdown_values.py:45]: Ahhh yes, unsure why this is here - but v clever that it gets added automatically. Does this apply to every query on `events`?

**Comment by Gilbert09** [comment]
@mariusandra I've added support for histogram breakdowns now with the latest commit: https://github.com/PostHog/posthog/pull/17891/commits/4b57daf06624c2321d60039ea849a91d3a5c9fd5

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown.py:53]:
```suggestion
        values = [f"[{t[0]},{t[1]}]" for t in buckets]
```

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown.py:34]: These two look inconsistent:

```
self._get_breakdown_buckets_ast()
self._get_breakdown_values_ast
```

I'd remove the `_get` from the property

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown.py:60]: I've usually seen `map` written as a inline list loop in Python:

```suggestion
        return ast.Array(exprs=[ast.Constant(value=v) for v in self._breakdown_values])
```

File: posthog/hogql_queries/insights/trends/breakdown_values.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown_values.py:82]: You can omit the timezone call and just write `timestamp`. A wrapper with the correct timezone will be added when the field is printed later.

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown.py:71]: Would be great to time and label this separately

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown.py:11]: Would be good to add return types to all the functions here, and makes our lives a bit easier later.

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown.py:42]: The name of this function confused me a bit. I would have likely called it something like `column_expr()` or something. Fine to keep, though 🤷

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown.py:83]: Why the `+ 0.01`? Won't that make some overlapping segments?

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown.py:116]: should the `["",""]` be `[0,0]` here?

File: posthog/hogql_queries/insights/trends/breakdown_values.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/breakdown_values.py:102]: Is this `.2f` precision enough for a large bin count?

File: posthog/hogql_queries/insights/trends/query_builder.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/query_builder.py:177]: Same comment re timezones

File: posthog/hogql_queries/insights/trends/query_builder.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/query_builder.py:208]: Hm... won't this just exclude everything that's not under the top N breakdown options? Is this how the current query worked? I'd assume we'd like everything else to be returned under a "other" blob? 🤔

File: posthog/hogql_queries/insights/trends/trends_query_runner.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/trends_query_runner.py:57]: The parent query runner returns just one `ast.SelectQuery` in the `to_query` method. This should power the "to hogql" button next to insights in the interface. Returning multiple queries breaks that flow.

Is there some way to return one query from here, even if just a large `union all`... and the runner will keep running individual queries?

Also, one other consideration: when we move to CH Cloud, we'll have access to a lot of parallelisation. This runner will run each query serially. Not blocking for now, but maybe we'll want to move the merge and formula into clickhouse somehow.

File: posthog/hogql_queries/insights/trends/trends_query_runner.py
Comment (mariusandra) [inline posthog/hogql_queries/insights/trends/trends_query_runner.py:218]: Some teams, like ours, have millions of event properties due to old plugins. Maybe we don't want all of them in memory 😬 

image

**Review by mariusandra — COMMENTED** [review]
All right, finally got time for a proper look. Left comments below 👍

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (Gilbert09) [reply posthog/hogql_queries/insights/trends/breakdown.py:71]: Could you expand a bit on what you mean here?

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (Gilbert09) [reply posthog/hogql_queries/insights/trends/breakdown.py:83]: Have added a comment to explain this, but it's taken from the current implementation:
```python
# Since we always `floor(x, 2)` the value, we add 0.01 to the last bucket
# to ensure it's always slightly greater than the maximum value
```

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (Gilbert09) [reply posthog/hogql_queries/insights/trends/breakdown.py:116]: The current implementation is set to just empty strings - I agree it's not ideal, but keen to change this post-release once we're at parity

File: posthog/hogql_queries/insights/trends/breakdown_values.py
Comment (Gilbert09) [reply posthog/hogql_queries/insights/trends/breakdown_values.py:102]: Taken from the current implementation, seems like it's been good enough in the past 🤷

File: posthog/hogql_queries/insights/trends/query_builder.py
Comment (Gilbert09) [reply posthog/hogql_queries/insights/trends/query_builder.py:208]: The current implementation only takes the top 25 results, ordered by `count()` at the moment. No bucketing via an "other" blob

File: posthog/hogql_queries/insights/trends/trends_query_runner.py
Comment (Gilbert09) [reply posthog/hogql_queries/insights/trends/trends_query_runner.py:57]: Would be keen to change this in an upcoming PR - can add it to my list of things TODO. It's been like this from the beginning of the trends query runner.

I'm expecting some insights won't run as a `UNION ALL` from the "to hogql" button due to the size of the query itself. Imagine 25 x of these queries in a single union all. If we want the `to_query` method to return a single query, then cool, but we'll likely need to break it down into each `SELECT` before passing them to the `execute` method

File: posthog/hogql_queries/insights/trends/trends_query_runner.py
Comment (Gilbert09) [reply posthog/hogql_queries/insights/trends/trends_query_runner.py:218]: Yep, that's fair - have changed this to a `get` on the individual field as opposed to `get all and then filter`

**Comment by Gilbert09** [comment]
@mariusandra PR fixes have been pushed - ready for a second review 🥳

File: posthog/hogql_queries/insights/trends/breakdown.py
Comment (mariusandra) [reply posthog/hogql_queries/insights/trends/breakdown.py]: ah, I meant that this query is making an internal query, so something like `with self.timings.measure('do the breakdown dance')` around it would make sense

File: posthog/hogql_queries/insights/trends/trends_query_runner.py
Comment (mariusandra) [reply posthog/hogql_queries/insights/trends/trends_query_runner.py:57]: Yep, but ClickHouse Cloud should ideally then parallelise those 25 queries and get the results much faster than us doing the loop in Python ever will.

**Review by mariusandra — APPROVED** [review]
LGTM

**Pull request merged**
